c10/core/TensorImpl.h

```diff
@@ -127,7 +127,13 @@ struct C10_API PlacementDeleteContext {
   }
 };
 
+struct TensorImpl;
+
 struct C10_API AutogradMetaInterface {
+  virtual void set_requires_grad(bool requires_grad, at::TensorImpl* self_impl) = 0;
+  virtual bool requires_grad() const = 0;
+  virtual at::Tensor& grad() = 0;
+  virtual const at::Tensor& grad() const = 0;
```

Comment:
We add these functions into `AutogradMetaInterface`, and call them from TensorImpl's corresponding functions, because Variable::Impl will not be overriding those functions anymore (and Variable::Impl will be removed in the next PR).